src/cmdline.rs

```diff
@@ -71,13 +72,48 @@ fn usage() -> Command {
 /// Parse the commandline into a `Command` to execute.
 pub fn parse() -> Command {
     trace!("parse");
-    let matches = get_app().get_matches();
-
+    let cwd = if let Ok(cwd) = env::current_dir() {
+        cwd
+    } else {
+        println!("sccache: Couldn't determine current working directory");
+        return usage();
+    };
     // The internal start server command is passed in the environment.
     let internal_start_server = match env::var("SCCACHE_START_SERVER") {
         Ok(val) => val == "1",
         Err(_) => false,
     };
+    let mut args: Vec<_> = env::args_os().collect();
+    if ! internal_start_server {
+        if let Ok(exe) = env::current_exe() {
+            match exe.file_stem().and_then(|s| s.to_str()).map(|s| s.to_lowercase()) {
+                // If the executable has its standard name, do nothing.
+                Some(ref e) if e == env!("CARGO_PKG_NAME") => {}
+                // Otherwise, if it was copied/hardlinked under a different $name, act
+                // as if it were invoked with `sccache $name`, but avoid $name resolving
+                // to ourselves again if it's in the PATH.
+                _ => {
+                    let path = env::var_os("PATH");
+                    match which_in(exe.file_name().unwrap(), path.as_ref(), &cwd) {
```

Comment:
I know most of these `unwrap`s are unlikely to ever be a problem, but they do rub me the wrong way a bit. What if you instead made `parse` return a `Result<Command>`, and then tweaked main to accommodate and print errors? Then you could basically write the same code but with `try!`. You might need to create a new `Error` type with conversions from various other `Error`s for this to work properly. I've seen `error-chain` recommended for that, although I haven't used it: https://github.com/brson/error-chain